src/librustc_trans/trans/consts.rs

```diff
@@ -114,29 +107,87 @@ pub fn trans_constval<'blk, 'tcx>(bcx: common::Block<'blk, 'tcx>,
         ConstVal::Str(ref v) => C_str_slice(ccx, v.clone()),
         ConstVal::ByteStr(ref v) => addr_of(ccx, C_bytes(ccx, v), 1, "byte_str"),
-        ConstVal::Struct(id) | ConstVal::Tuple(id) => {
-            let expr = bcx.tcx().map.expect_expr(id);
-            match const_expr(ccx, expr, param_substs, None, TrueConst::Yes) {
-                Ok((val, _)) => val,
-                Err(e) => panic!("const eval failure: {}", e.description()),
+        ConstVal::Struct(id, ref field_values) => {
+            let repr = adt::represent_type(ccx, ty);
+            let mut trans_fields = Vec::with_capacity(field_values.len());
+            let VariantInfo { discr, fields } = VariantInfo::of_node(ccx.tcx(), ty, id);
+            for Field(f_name, f_ty) in fields {
+                let value = field_values.get(&f_name)
+                                        .expect("trans knows struct fields that const doesn't");
+                let value = trans_constval(bcx, value, f_ty, param_substs);
+                trans_fields.push(value);
+            }
+            // FIXME: check that all elements of `field_values` have been translated
+            if ty.is_simd() {
+                C_vector(&trans_fields[..])
+            } else {
+                adt::trans_const(ccx, &*repr, discr, &trans_fields)
             }
         },
-        ConstVal::Array(id, _) | ConstVal::Repeat(id, _) => {
-            let expr = bcx.tcx().map.expect_expr(id);
-            expr::trans(bcx, expr).datum.val
+        ConstVal::Tuple(id, ref field_values) => {
+            let repr = adt::represent_type(ccx, ty);
+            let fields = VariantInfo::of_node(ccx.tcx(), ty, id)
+                .fields
+                .into_iter()
+                .map(|Field(_, f_ty)| f_ty)
+                .zip(field_values)
+                .map(|(f_ty, f_val)| trans_constval(bcx, f_val, f_ty, param_substs))
+                .collect::<Vec<_>>();
+            adt::trans_const(ccx, &*repr, Disr(0), &fields)
+        },
+        ConstVal::Array(_, _, ref values) => {
+            let unit_ty = ty.sequence_element_type(ccx.tcx());
+            let llunitty = type_of::type_of(ccx, unit_ty);
+            let vs = values.iter()
+                           .map(|val| trans_constval(bcx, val, unit_ty, param_substs))
+                           .collect::<Vec<_>>();
+            // If the vector contains enums, an LLVM array won't work.
+            if vs.iter().any(|vi| val_ty(*vi) != llunitty) {
+                C_struct(ccx, &vs, false)
+            } else {
+                C_array(llunitty, &vs)
+            }
+        },
+        ConstVal::Repeat(_, n, ref val) => {
+            let unit_ty = ty.sequence_element_type(ccx.tcx());
+            let llunitty = type_of::type_of(ccx, unit_ty);
+            let unit_val = trans_constval(bcx, val, unit_ty, param_substs);
+            let vs = vec![unit_val; n as usize];
```

Comment:
jup :) how else would you allow someone to index into it without knowing it's a Repeat expression? If it is indexed in the const evaluator, this is a non-issue anyway, since that just does a range check and accesses the value.